File: src/go/entityfx/scimark2/scimark2.go
package scimark2

import (
	"../utils"
)

type Scimark2Result struct {
	compositeScore float64
	sor float64
	fft float64
	monteCarlo float64
	sparseMathmult float64
	lu float64
}

func Bench(min_time float64, isLarge bool, output utils.WriterType) *Scimark2Result {
	fFT_size := FFT_SIZE
	sOR_size := SOR_SIZE
	sparse_size_M := SPARSE_SIZE_M
	sparse_size_nz := SPARSE_SIZE_nz
	lU_size := LU_SIZE
	current_arg := 0
	if isLarge {
		fFT_size = LG_FFT_SIZE
		sOR_size = LG_SOR_SIZE
		sparse_size_M = LG_SPARSE_SIZE_M
		sparse_size_nz = LG_SPARSE_SIZE_nz
		lU_size = LG_LU_SIZE
		current_arg++
	}

	res := [6]float64{}

	res[1] = measureFFT(fFT_size, min_time)
	res[2] = measureSOR(sOR_size, min_time)
	res[3] = measureMonteCarlo(min_time)
	res[4] = measureSparseMatmult(sparse_size_M, sparse_size_nz, min_time)
	res[5] = measureLU(lU_size, min_time)
	res[0] = (((res[1] + res[2] + res[3]) + res[4] + res[5])) / (5.0)

	output.WriteNewLine() 
	output.WriteLine("SciMark 2.0a") 
	output.WriteNewLine() 
	output.WriteLine("Composite Score: %.2f", res[0]) 
	output.Write("FFT (%d): ", fFT_size) 
	if (res[1] == .0) {
		output.WriteLine(" ERROR, INVALID NUMERICAL RESULT!")
	} else {
		output.WriteLine("%.2f", res[1]) 
	}
	output.WriteLine("SOR (%dx%d):   %.2f", sOR_size, sOR_size, res[2]) 
	output.WriteLine("Monte Carlo : %.2f", res[3]) 
	output.WriteLine("Sparse matmult (N=%d, nz=%d): %.2f", sparse_size_M, sparse_size_nz, res[4]) 
	output.Write("LU (%dx%d): ", lU_size, lU_size) 
	if (res[5] == .0) {
		output.WriteLine(" ERROR, INVALID NUMERICAL RESULT!") 
	} else {
		output.WriteLine("%.2f", res[5]) 
	}
	return &Scimark2Result{res[0], res[1], res[2], res[3], res[4], res[5]}
}

File: src/dotnet/EntityFX.NetBenchamarks.Core/Scimark2/Scimark2.cs
using System;

namespace EntityFX.NetBenchmark.Core.Scimark2
{

    public class Scimark2
    {
        IWriter output;

        public Scimark2(bool printToConsole)
        {
            output = new Writer(null);
            output.UseConsole = printToConsole;
        }

        public Scimark2Result Bench(double min_time, bool isLarge)
        {
            // default to the (small) cache-contained version

            int FFT_size = Constants.FFT_SIZE;
            int SOR_size = Constants.SOR_SIZE;
            int Sparse_size_M = Constants.SPARSE_SIZE_M;
            int Sparse_size_nz = Constants.SPARSE_SIZE_nz;
            int LU_size = Constants.LU_SIZE;

            // look for runtime options

            int current_arg = 0;
            if (isLarge)
            {
                FFT_size = Constants.LG_FFT_SIZE;
                SOR_size = Constants.LG_SOR_SIZE;
                Sparse_size_M = Constants.LG_SPARSE_SIZE_M;
                Sparse_size_nz = Constants.LG_SPARSE_SIZE_nz;
                LU_size = Constants.LG_LU_SIZE;

                current_arg++;
            }


            // run the benchmark

            double[] res = new double[6];
            Random R = new Random(Constants.RANDOM_SEED);

            res[1] = kernel.measureFFT(FFT_size, min_time, R);
            res[2] = kernel.measureSOR(SOR_size, min_time, R);
            res[3] = kernel.measureMonteCarlo(min_time, R);
            res[4] = kernel.measureSparseMatmult(Sparse_size_M,
                        Sparse_size_nz, min_time, R);
            res[5] = kernel.measureLU(LU_size, min_time, R);


            res[0] = (res[1] + res[2] + res[3] + res[4] + res[5]) / 5;


            // print out results

            output.WriteLine();
            output.WriteLine("SciMark 2.0a");
            output.WriteLine();
            output.WriteLine("Composite Score: {0}", res[0]);
            output.Write("FFT ({0}): ", FFT_size);
            if (res[1] == 0.0)
                output.WriteLine(" ERROR, INVALID NUMERICAL RESULT!");

            else
                output.WriteLine("{0}", res[1]);

            output.WriteLine("SOR ({0}x{1}):   {2}", SOR_size, SOR_size, res[2]);
            output.WriteLine("Monte Carlo : {0}", res[3]);
            output.WriteLine("Sparse matmult (N={0}, nz={1}): {2}", Sparse_size_M, Sparse_size_nz, res[4]);
            output.Write("LU ({0}x{1}): ", LU_size, LU_size);
            if (res[5] == 0.0)
                output.WriteLine(" ERROR, INVALID NUMERICAL RESULT!");

            else
                output.WriteLine("{0}", res[5]);

            return new Scimark2Result()
            {
                CompositeScore = res[0],
                FFT = res[1],
                SOR = res[2],
                MonteCarlo = res[3],
                SparseMathmult = res[4],
                LU = res[5],
                Output = output.Output
            };
        }

    }
}